core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java

```diff
@@ -321,7 +321,7 @@ private void handleResponseError(CompletionStage<ValidResponse> rpcFuture, Objec
             log.tracef("UnRegistered change listener for %s", requestId);
          }
       }
-      if (cache.getStatus() != ComponentStatus.RUNNING) {
+      if (cache.getStatus() != ComponentStatus.RUNNING && cache.getStatus() != ComponentStatus.INITIALIZING) {
```

Comment:
Actually now that I think about this more, is INITIALIZING an okay state? Could the rehash still be in progress?

Reply:
Yes, a rebalance could be in progress, but a rebalance can also be in progress in the `RUNNING` state so TBH I'm not sure what this is guarding against.

Reply:
I found in testing that when a node is shutting down it can clear the data container yet still return a response (this led to commands that have a subset of data). The more I think about it, the real fix is probably to ensure that a node can't send a command response before clearing the data container on shut down.

Reply:
This should take care of not sending successful responses after clearing the data container: https://github.com/infinispan/infinispan/blob/a72f53962a68f7e38fc8a472144c878be35f3d46/core/src/main/java/org/infinispan/remoting/inboundhandler/BaseBlockingRunnable.java#L118

True, it checks the `PerCacheInboundInvocationHandler` status, not the cache status, but the data container clear should be the last thing to happen during shutdown. Maybe in your failed test the data was in a store, and `PersistenceManagerImpl.stop()` is running before `PerCacheInboundInvocationHandler.stop()`?

Reply:
The test is in memory only, I can try running the test without these lines. However there is another bug somewhere I am trying to track down first. I will definitely try this out though.

Reply:
Ok, should I open a PR with just the fix for `getLocalReadSegments()`?

Reply:
I created https://issues.jboss.org/browse/ISPN-9542 already which has the fix. I also added an optimization for REPL so it doesn't retry due to rehash anymore. :)

There is still a bug in DIST I can't figure out yet, been trying to isolate it.

Reply:
So it fails with these lines commented out every time. The issue isn't that the cache is stopped, but rather that it is running, accepts the command and then stops while in the middle of processing it.

Reply:
And the reason it doesn't use the `BaseBlockingRunnable` check is because it sends its own response manually. https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java#L406